wal/src/manager.rs

```diff
@@ -255,10 +266,13 @@ pub trait WalManager: Send + Sync + fmt::Debug {
         req: &ReadRequest,
     ) -> Result<BatchLogIteratorAdapter>;
 
+    /// Provide the encoder for encoding payloads.
+    async fn encoder(&self, region_id: RegionId, entries_num: u64) -> Result<LogBatchEncoder>;
```

Comment:
Yes, it will brings some concurrent problems if thread model change.

This give me a new idea about `pre-code`:

- Because `SequenceNumber` is only used in `Encoding key`, maybe we can move encoding `key` to write method. Finally, it will never has concurrent problems, because `allocates SequenceNumber` and write is both in the same write lock.

Encode key:
```rust
// pub type LogKey = (RegionId, SequenceNumber);
pub fn encode_key(&self, buf: &mut BytesMut, log_key: &LogKey) -> Result<()>;
```

Encode Value:
```rust
// Never use SequenceNumber
pub fn encode_value(&self, buf: &mut BytesMut, payload: &impl Payload) -> Result<()>
```

Region in kv_table_impl:
```rust
pub struct Region {
    runtimes: WalRuntimes,
    state: RegionState,
    writer: Mutex<RegionWriter>,
}
```

Reply:
Sounds good. `LogKey` is fixed-length, payload-unrelated metadata, we don't have to encode both of them at the same time.

> because allocates SequenceNumber and write are in the protect of same mutex.

For now `SeqNum` is an atomic but yes, we can put "allocate `SeqNum`" and "consume(write) `SeqNum`)" together. The "mutex" is our thread model, let's leave it as it is.

Reply:
Addressed.